### PR TITLE
[cmd/opampsupervisor]: Persist the instance ID between restarts

### LINKED ISSUE #21073: [cmd/opampsupervisor] Persist Collector instance ID
### Component(s)

_No response_

### Is your feature request related to a problem? Please describe.

The Supervisor should create Collector instance ID to set in the Collector's resource attributes, and should also persist this ID so that it stays the same across Supervisor restarts. For more, see the [design document](https://docs.google.com/document/d/1KtH5atZQUs9Achbce6LiOaJxLbksNJenvgvyKLsJrkc/edit#heading=h.8szgxmjard0e).

### Describe the solution you'd like

Allow configuring where the Supervisor persists information according to the config schema defined in the design document and store the instance ID in this location.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by BinaryFissionGames**
**Description:** <Describe what has changed.>
* Adds the ability to persist the instance ID on disk and load it on startup.

**Link to tracking Issue:** Closes #21073

**Testing:**
* Add unit tests for persistence functionality
* Add e2e tests for persistence (testing that initial id is persisted, and that server specified ID is persisted)

**Documentation:** N/A

### CODE REVIEW

File: cmd/opampsupervisor/supervisor/persistence.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+)
+
+// persistentState represents persistent state for the supervisor
+type persistentState struct {
+	InstanceID ulid.ULID `yaml:"instance_id"`
+
+	// Path to the config file that the state should be saved to.
+	// This is not marshaled.
+	configPath string `yaml:"-"`
+}
```
Comment (BinaryFissionGames) [inline cmd/opampsupervisor/supervisor/persistence.go:23]: This might be overkill to have this be a whole struct. We could get away with just writing the instance ID directly to a file instead of as marshaled YAML.

The advantage of doing it this way is that we can extend in the future to include more supervisor state if we need to add anything here. We can also version this state in the future if there are any changes to it (e.g. port the ULID representation over to arbitrary bytes representation, if we end up making that change)

File: cmd/opampsupervisor/supervisor/config/config.go [hunk 2]
```diff
@@ -14,6 +17,35 @@ type Supervisor struct {
...
+			}
+			return filepath.Join(programDataDir, "Otelcol", "Supervisor")
+		default:
+			// Default for non-windows systems
+			return "/var/lib/otelcol/supervisor"
+		}
+	}
+
+	return s.Directory
 }
```
Comment (BinaryFissionGames) [inline cmd/opampsupervisor/supervisor/config/config.go:49]: Might make sense to move this to a separate PR, I think we'll need to create these directories if they don't exist too.

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 3]
```diff
@@ -826,10 +815,18 @@ func (s *Supervisor) runAgentProcess() {
...
 
 		case <-s.healthCheckTicker.C:
 			s.healthCheck()
+
+		case <-s.doneChan:
+			err := s.commander.Stop(context.Background())
+			if err != nil {
+				s.logger.Error("Could not stop agent process", zap.Error(err))
+			}
+			return
```
Comment (BinaryFissionGames) [inline cmd/opampsupervisor/supervisor/supervisor.go:829]: This change is due to a race condition I was bumping into with the e2e tests. Basically, the shutdown would happen while the command was in the middle of starting, so this goroutine would never exit (the command never exited), then Supervisor.Shutdown would never exit as a consequence. 

Instead, we just handle the full lifecycle of the command all in the same goroutine.

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 4]
```diff
@@ -889,11 +879,11 @@ func (s *Supervisor) Shutdown() {
 		}
 	}
 
+	s.supervisorWG.Wait()
+
 	if s.healthCheckTicker != nil {
 		s.healthCheckTicker.Stop()
 	}
```
Comment (BinaryFissionGames) [inline cmd/opampsupervisor/supervisor/supervisor.go:886]: This was another race condition, the health check ticker is recreated by the agent command goroutine, so we need to make sure that goroutine is exited before we try to stop it (otherwise we may leak a non-stopped ticker)

**Comment by BinaryFissionGames** [comment]
@evan-bradley It's ready for review now, I tend to open things in draft and come back to them with a fresh mind to look back over the PR, turns out it was a good idea because I caught a few weird naming things here 😆 

I do think there's a small bit of overlap in #30807 - but I think it's only in the configuration. I think it makes sense to persist those statuses in different files, just because they are raw byte payloads that probably wouldn't be so useful in a structured yaml file like this.

File: cmd/opampsupervisor/supervisor/supervisor.go [hunk 5]
```diff
@@ -955,6 +950,10 @@ func (s *Supervisor) onMessage(ctx context.Context, msg *types.MessageData) {
...
 }
 
+func (s *Supervisor) persistentStateFile() string {
+	return filepath.Join(s.config.Storage.DirectoryOrDefault(), persistentStateFile)
```
Comment (tigrannajaryan) [inline cmd/opampsupervisor/supervisor/supervisor.go:954]: Should we `mkdir` the directory in case it does not exist?

Comment (BinaryFissionGames) [reply cmd/opampsupervisor/supervisor/supervisor.go:1054]: Yeah, we should. I've added the directory creation logic in the `NewSupervisor` function.

File: cmd/opampsupervisor/supervisor/persistence.go (hunk 1 above)
Comment (tigrannajaryan) [reply cmd/opampsupervisor/supervisor/persistence.go:23]: +1 to yaml.

File: cmd/opampsupervisor/supervisor/config/config.go [hunk 6]
```diff
@@ -14,14 +17,37 @@ type Supervisor struct {
...
 
 type Storage struct {
 	// Directory is the directory where the Supervisor will store its data.
 	Directory string `mapstructure:"directory"`
 }
 
+// DirectoryOrDefault returns the configured storage directory if it was configured,
+// otherwise it returns the system default.
+func (s Storage) DirectoryOrDefault() string {
```
[the remark this answers is not in the document]
Comment (BinaryFissionGames) [reply cmd/opampsupervisor/supervisor/config/config.go:30]: Added a section to the README explaining this; It's in the spec, but it makes sense that we'll have to work on moving some of that stuff into the README as we implement the spec.

**Review by evan-bradley — APPROVED** [review]

**Comment by bogdandrutu** [comment]
@tigrannajaryan is this ready to be merged from your perspective?

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan is this ready to be merged from your perspective?

User-visible functionality LGTM. I didn't review the code, but @evan-bradley did so I believe it should be good to merge.
